### PR TITLE
[libbeat] AWS ELB Monitoring

### PR DESCRIPTION
**Pull request by andrewvc**
Autodiscovery provider for AWS ELBs.

Not a lot to add besides that, the rest is in the docs provided within this PR.
 
One open question is the amount of testing to do here. We're still not sure if we want to introduce this as a beta feature, which might change the answer there. Ultimately, the AWS library is very tricky to mock. There is an elasticloadbalancingv2iface package that supposedly makes things easier, but realistically, you have to implement so much the value of that is dubious. Real functional testing with the AWS APIs is probably the only satisfactory answer, but that too is a significant investment.

### CODE REVIEW

**Comment by exekias** [comment]
:see_no_evil: :smile: interesting, we are currently talking about adding EC2 autodiscover provider! cc @kaiyan-sheng

**Comment by ruflin** [comment]
I suggest we merge the first PR as experimental and do some testing on it for the next minor release. If we get comfortable with it we can push it very quickly into beta (or even GA).

Didn't check the code yet in detail but for sure it will need some docs.

**Comment by dov0211** [comment]
+1 on releasing it as an experimental feature

**Comment by andrewvc** [comment]
Per discussion with @kaiyan-sheng and @exekias we should ensure that the config here is consistent with the EC2 module. We also need to add the session token option as part of that.

File: libbeat/docs/shared-autodiscover.asciidoc [hunk 1]
```diff
@@ -257,6 +257,39 @@ the ones used for discovery probes, they have these settings:
...
+===== Amazon ELBs
+
+The Amazon ELB autodiscover provider discovers ELBs and their listeners. This is useful when you don't want to connect
+directly to a service, but rather to the ELB fronting a pool of services.
+
+This provider will yield one config block per ELB Listener. So, if you have one ELB exposing both ports 80 and 443, it
+will generate two configs, one for each port. Keep in mind that the beat will de-duplicate configs. So, if the generated
+configs are the same only one will actually run.
+
+This provider will load AWS credentials using the standard AWS environment variables and shared credentials files see https://docs.aws.amazon.com/general/latest/gr/aws-access-keys-best-practices.html[Best Practices for Managing AWS Access Keys] for more information. If you do not wish to use these, you may explicitly set the `credentials.access_key_id` and `credentials.secret_access_key` variables.
```
Comment (kaiyan-sheng) [inline libbeat/docs/shared-autodiscover.asciidoc]: Based on https://github.com/elastic/beats/pull/12401/files#diff-f0adf3363109da68d4f12705a9541b95R11, why it has `credentials.*` instead of just `access_key_id`?

File: x-pack/libbeat/autodiscover/providers/aws/elb/config.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	Builders     []*common.Config        `config:"builders"`
+	Appenders    []*common.Config        `config:"appenders"`
+	Templates    template.MapperSettings `config:"templates"`
+
+	// Period defines how often to poll the AWS API.
+	Period time.Duration `config:"period" validate:"nonzero,required"`
+
+	// AWS Specific autodiscover fields
+
+	Regions         []string `config:"regions" validate:"required"`
```
Comment (kaiyan-sheng) [inline x-pack/libbeat/autodiscover/providers/aws/elb/config.go]: In aws metricbeat module, we allow user to not set `Regions`. In that case, we will go query for all regions and loop through them. This might not be a good practice here but I just want to point it out.

File: libbeat/docs/shared-autodiscover.asciidoc [hunk 3]
```diff
@@ -257,6 +257,39 @@ the ones used for discovery probes, they have these settings:
...
+
+These are the available fields on every event
+
+  * host
+  * port
+  * elb_listener.listener_arn
```
Comment (ruflin) [inline libbeat/docs/shared-autodiscover.asciidoc]: Do we need all these fields? Are there use cases for `ssl_policy` for example?

File: vendor/vendor.json [hunk 4]
```diff
@@ -168,10 +168,10 @@
 			"versionExact": "v2.0.0-preview.5"
 		},
 		{
-			"checksumSHA1": "1AyIehd0sjJ9uJoLZyqq9L9a1Jg=",
+			"checksumSHA1": "427ei4bQ6ap8N/Hrw7NpxOqjd+Y=",
```
Comment (ruflin) [inline vendor/vendor.json]: I wonder if it would be worth doing a separate PR to just update to this dependency, making sure it has not side effects. And then add the two additional packages in this PR.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  title: "ELB Listener"
+  description: >
+    ELB Listeners
+  short_config: false
+  release: experimental
+  fields:
+    - name: elb_listener
+      type: group
+      description: >
+          jolokia contains metrics exposed via jolokia agent
```
Comment (ruflin) [inline x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: ?

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml [hunk 6]
```diff
@@ -0,0 +1,84 @@
+- key: elb_listener
```
Comment (ruflin) [inline x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: Should wecall it just `elb`?

File: libbeat/docs/shared-autodiscover.asciidoc [hunk 7]
```diff
@@ -257,6 +257,39 @@ the ones used for discovery probes, they have these settings:
...
+
+This provider will yield one config block per ELB Listener. So, if you have one ELB exposing both ports 80 and 443, it
+will generate two configs, one for each port. Keep in mind that the beat will de-duplicate configs. So, if the generated
+configs are the same only one will actually run.
+
+This provider will load AWS credentials using the standard AWS environment variables and shared credentials files see https://docs.aws.amazon.com/general/latest/gr/aws-access-keys-best-practices.html[Best Practices for Managing AWS Access Keys] for more information. If you do not wish to use these, you may explicitly set the `credentials.access_key_id` and `credentials.secret_access_key` variables.
+
+These are the available fields on every event
+
+  * host
```
Comment (ruflin) [inline libbeat/docs/shared-autodiscover.asciidoc]: Not directly but I thought worth mentioning: This is not following ECS but is the same in other autodiscovery providers. I wonder if long term this could become an issue if we would also ingest some autodiscovery data for later querying.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  release: experimental
+  fields:
+    - name: elb_listener
+      type: group
+      description: >
+          jolokia contains metrics exposed via jolokia agent
+      fields:
+        - name: ip_address_type
+          type: keyword
+          level: core
```
Comment (ruflin) [inline x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: Where is this level part come from? We only have this in ECS AFAIK.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml [hunk 9]
```diff
@@ -0,0 +1,84 @@
+- key: elb_listener
+  title: "ELB Listener"
+  description: >
+    ELB Listeners
+  short_config: false
+  release: experimental
+  fields:
+    - name: elb_listener
```
Comment (ruflin) [inline x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: Should we try to map some of the fields below to ECS? Would that be more confusing then valuable? Undecided.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+          type: date
+          level: core
+          example: 2019-06-12T21:55:14.490Z
+          description: Date of creation
+        - name: availability_zones
+          type: keyword
+          level: core
+          example: "us-east-1e"
+          description: List of AZs this is active in
+        - name: host
```
Comment (ruflin) [inline x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: In the docs you listed `host` on the top level but here it is nested. Same for port, is this correct?

File: x-pack/libbeat/autodiscover/providers/aws/elb/lblistener.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+// toMap converts this lbListener into the form consumed as metadata in the autodiscovery process.
+func (l *lbListener) toMap() common.MapStr {
+	m := common.MapStr{}
+
+	// We fully spell out listener_arn to avoid confusion with the ARN for the whole ELB
+	m["listener_arn"] = l.listener.ListenerArn
+	m["load_balancer_arn"] = *l.lb.LoadBalancerArn
+	m["host"] = *l.lb.DNSName
```
Comment (ruflin) [inline x-pack/libbeat/autodiscover/providers/aws/elb/lblistener.go]: host is inside here, I wonder if the docs and example are correct.

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2/elasticloadbalancingv2iface"
+
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+
+	"github.com/gofrs/uuid"
+
+	"github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2"
+
```
Comment (ruflin) [inline x-pack/libbeat/autodiscover/providers/aws/elb/provider.go]: Nit: Import sorting fun :-)

File: libbeat/docs/shared-autodiscover.asciidoc (hunk 1 above)
Comment (andrewvc) [reply libbeat/docs/shared-autodiscover.asciidoc]: Good catch, that was from an earlier revision. It's the docs here that are wrong.

File: libbeat/docs/shared-autodiscover.asciidoc (hunk 3 above)
Comment (andrewvc) [reply libbeat/docs/shared-autodiscover.asciidoc]: The ARNs are really important for any sort of programatic stuff. The SSL policy could possibly be dispensed with, but I'd argue that every other field is useful, so at that point, I'd say we should keep it for completeness.

File: libbeat/docs/shared-autodiscover.asciidoc (hunk 7 above)
Comment (andrewvc) [reply libbeat/docs/shared-autodiscover.asciidoc]: The `host` and `port` fields aren't inserted into the event. Only the `elb_listener` fields are. The fields listed here are what are present in the templating phase of the config. I'll add info here to clarify.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml (hunk 8 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: Ohhhh, I was just blindly copy pasting ;)

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml (hunk 9 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: I think we should wait. I'm not sure if we're at the point where we want every AWS API mapped into ECS. EC2, sure. But ELBs feel too specific.

Would be curious to here from @MikePaquette and @webmat.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml (hunk 10 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: Yes, different host variable. The top level one isn't in the event, it's only available during config templating. That's a standard thing for autodiscover providers.

It's debatable whether we need to duplicate host here since it's part of `url.domain`, however it's part of the API response, and it feels weird leaving it incomplete.

**Comment by andrewvc** [comment]
@ruflin @kaiyan-sheng thanks for the reviews! I've incorporated all changes requested. Thanks for catching what you've caught. I've also formally moved this into review status.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml (hunk 6 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: An ELB is the load balancer, but each port it exposes is an ELB Listener. 

This provider actually emits ELB Listeners, because presumable you want to monitor each port, not each ELB. Orienting it around the listeners means you get all the right metadata like port etc.

File: x-pack/libbeat/autodiscover/providers/aws/elb/config.go (hunk 2 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/config.go]: Thanks for the heads up. Normally I'd say we should not change behavior, however, maybe we should make this a requirement going forward. There are a lot of AWS regions, and it seems like a waste. Not to mention, the user may only have perms for one region. Looping through them all may create unnecessary errors.

WDYT?

File: vendor/vendor.json (hunk 4 above)
Comment (andrewvc) [reply vendor/vendor.json]: I don't mind that. But I'm indifferent. Shall I do that?

**Comment by andrewvc** [comment]
@ruflin in reference to your question about API query costs, I couldn't find any costs. The pricing page for ELBs does not list them: https://aws.amazon.com/elasticloadbalancing/pricing/

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml (hunk 9 above)
Comment (webmat) [reply x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: Most of the stuff here is very AWS-specific (SGs, VPC id, etc). So most of them are fine as custom fields, IMO. Even the stuff that's kind of network-centric follows different semantics (`ip_address_type` and `type`), and I don't think we should try to fit those in ECS.

I do think it would make sense to at fill out the following, though:

* `destination.port`
* `destination.address` (from `host`)
* `destination.ip`

If Heatbeat knows its own IP and outgoing port, perhaps fill `source.[ip|port]` as well.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: Great, done!

File: vendor/vendor.json (hunk 4 above)
Comment (andrewvc) [reply vendor/vendor.json]: OK, having tried to fix the deps, I'll definitely do that tomorrow. Looks like the latest libs don't make lambdabeat happy.

File: x-pack/libbeat/autodiscover/providers/aws/elb/config.go (hunk 2 above)
Comment (kaiyan-sheng) [reply x-pack/libbeat/autodiscover/providers/aws/elb/config.go]: Yeah good point on the permission problem!

File: vendor/vendor.json (hunk 4 above)
Comment (ruflin) [reply vendor/vendor.json]: 👍

File: libbeat/docs/shared-autodiscover.asciidoc (hunk 7 above)
Comment (ruflin) [reply libbeat/docs/shared-autodiscover.asciidoc]: I'm aware that it's not added to the event, but I'm still thinking if we should potentially align to prevent confusion? @exekias ?

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider.go [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+func (p *Provider) onWatcherStart(arn string, lbl *lbListener) {
+	lblMap := lbl.toMap()
+	e := bus.Event{
+		"start":    true,
+		"provider": p.uuid,
+		"id":       arn,
+		"host":     lblMap["host"],
+		"port":     lblMap["port"],
+		"meta": common.MapStr{
+			"elb_listener": lbl.toMap(),
```
Comment (exekias) [inline x-pack/libbeat/autodiscover/providers/aws/elb/provider.go]: it would probably make sense to offer the same fields offered by `add_cloud_metadata` here, WDYT? I see some of them are under `elb_listener` already.

File: x-pack/libbeat/autodiscover/providers/aws/elb/lblistener.go [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+	m["port"] = *l.listener.Port
+	m["protocol"] = l.listener.Protocol
+	m["type"] = string(l.lb.Type)
+	m["scheme"] = l.lb.Scheme
+	m["availability_zones"] = l.azStrings()
+	m["created"] = l.lb.CreatedTime
+	m["state"] = l.stateMap()
+	m["ip_address_type"] = string(l.lb.IpAddressType)
+	m["security_groups"] = l.lb.SecurityGroups
+	m["vpc_id"] = *l.lb.VpcId
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/lblistener.go]: We should check if any of these derefereced variables are nil, if they are for any reason this will panic. (Is it possible to have a load balancer without DNS name for example?)

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+	"testing"
+	"time"
+
+	"github.com/gofrs/uuid"
+
+	"github.com/pkg/errors"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/stretchr/testify/require"
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go]: Group imports.

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+		if tea.len() >= targetLen {
+			return
+		}
+		time.Sleep(time.Millisecond)
+	}
+
+	t.Fatalf("Timed out waiting for num events to be %d", targetLen)
+}
+
+func Test_internalBuilder(t *testing.T) {
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go]:
```suggestion
func TestInternalBuilder(t *testing.T) {
```

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+func (w *watcher) stop() {
+	close(w.done)
+}
+
+func (w *watcher) forever() {
+	for {
+		select {
+		case <-w.done:
+			w.ticker.Stop()
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: And `return`?

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+		}
+		w.lbListeners[arn] = w.gen
+	}
+
+	// ELBs not seen in the API request get deleted
+	for uuid, entryGen := range w.lbListeners {
+		if entryGen == oldGen {
+			if w.onStop != nil {
+				w.onStop(uuid)
+			}
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: And `delete(w.lbListeners, uuid)`?

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+		}
+	}
+
+	return nil
+}
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: If a lb changes, its `arn` changes too? If not we may need something to detect changes.

File: x-pack/libbeat/autodiscover/providers/aws/elb/config.go [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+	AccessKeyID     string   `config:"access_key_id"`
+	SecretAccessKey string   `config:"secret_access_key"`
+	SessionToken    string   `config:"session_token"`
+}
```
Comment (kaiyan-sheng) [inline x-pack/libbeat/autodiscover/providers/aws/elb/config.go]: Maybe we should add support for profile name here as well? https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html so when using `external.LoadDefaultAWSConfig`, users can specify a profile name.

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider.go [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+
+	config := defaultConfig()
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	var clients []elasticloadbalancingv2iface.ClientAPI
+	for _, region := range config.Regions {
+		awsCfg, err := external.LoadDefaultAWSConfig()
```
Comment (kaiyan-sheng) [inline x-pack/libbeat/autodiscover/providers/aws/elb/provider.go]: I think LoadDefaultAWSConfig can be outside this for loop because besides region is different for each `awsCfg`, the credentials will be the same.

File: libbeat/docs/shared-autodiscover.asciidoc (hunk 7 above)
Comment (andrewvc) [reply libbeat/docs/shared-autodiscover.asciidoc]: Sorry, I think there is some confusion here. Those fields *do* contain their own `host` and `port` fields. Additionally, if using heartbeat you'd probably use `url.port` and `url.domain` anyway.

File: x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml (hunk 9 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/_meta/fields.yml]: Well, since most of these ECS fields will be filled by heartbeat or whatever other beat uses these I'm averse to starting to add extra `destination`  fields, I think users would find it confusing. Additionally, I'd argue that `destination` isn't a great fit in the first place here.

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider.go (hunk 13 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/provider.go]: ++

File: x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+
+func (p *fetchRequest) fetchAllPages() {
+	// Keep fetching pages unless we're stopped OR there are no pages left
+	for p.running.Load() && p.fetchNextPage() {
+		logp.Debug("autodiscover-elb", "API page fetched")
+	}
+}
+
+func (p *fetchRequest) fetchNextPage() (more bool) {
+	success := p.paginator.Next(context.TODO())
```
[the remark this answers is not in the document]
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go]: Great idea, will swap out those things for a context.

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go (hunk 18 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: Good catch!

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go (hunk 19 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: No, it's ARN should be stable, the ARN is what identifies it, so by definition it cannot change.

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go (hunk 17 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: Good catch!

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go (hunk 16 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go]: This is actually the preferred pattern for naming method tests, inherited from godoc examples where underscore delineates a method vs. a plain function named `InternalBuilder`. See [testing](https://golang.org/pkg/testing/#hdr-Examples) for more info. 

I think the go team should be more clear here IMHO, and I realize that this usage is less common, but I think the underscore makes sense because it's more clear that `internalBuilder` is not exported, vs. capitalizing it.

File: x-pack/libbeat/autodiscover/providers/aws/elb/lblistener.go (hunk 14 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/lblistener.go]: I've made this safer. There is always a DNS name set by amazon. The AWS lib just always uses `*string` instead of `string` for reasons that are unclear to me 🤷‍♂ . I assume they think it's more efficient memory-wise.

**Comment by andrewvc** [comment]
@jsoriano thanks for the review! Apologies for the quite long delay. I think I've addressed all your points. Adding the context was a big when.

I also set a context timeout for the API fetch to be equal to the period between refreshes. I debated making this configurable by the user, but I don't think having this be tunable is practically helpful. 

WRT integration tests, I agree that those can wait till this is out of experimental status.

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go (hunk 16 above)
Comment (jsoriano) [reply x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go]: > See testing for more info.

The examples there are for example code, where some convention is actually needed to match certain objects to its examples in docs. There the underscore is reserved for struct methods, that would require a dot otherwise, what wouldn't be valid.
In general I'd prefer to don't use underscores, but I agree that these conventions are not clear enough for these cases, and the point about it being a not exported function is a good one, so as you prefer :slightly_smiling_face:

File: x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go [hunk 23]
```diff
@@ -136,13 +137,21 @@ func (p *fetchRequest) fetch() ([]*lbListener, error) {
...
 func (p *fetchRequest) fetchAllPages() {
 	// Keep fetching pages unless we're stopped OR there are no pages left
-	for p.running.Load() && p.fetchNextPage() {
-		logp.Debug("autodiscover-elb", "API page fetched")
+Outer:
+	for {
+		select {
+		case <-p.context.Done():
+			logp.Debug(logSelector, "done fetching ELB pages, context cancelled")
+			break Outer
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go]: Or just `return`?

File: x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go [hunk 24]
```diff
@@ -175,14 +184,23 @@ func (p *fetchRequest) dispatch(fn func()) {
...
+
 	if listen.Err() != nil {
 		p.recordErrResult(listen.Err())
 	}
+
+Outer:
+	for listen.Next(p.context) {
+		select {
+		case <-p.context.Done():
+			break Outer
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go]: Here we could also `return`. In any case, maybe `listen.Next` also returns `false` the passed context is done?

File: x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+
+package elb
+
+import (
+	"context"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2/elasticloadbalancingv2iface"
+
+	"github.com/elastic/beats/libbeat/logp"
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go]: Nit. Our own packages should be at the end on imports, this import block should be something like:
```
import (
	"context"
	"sync"

	"github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2"
	"github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2/elasticloadbalancingv2iface"
	"go.uber.org/multierr"

	"github.com/elastic/beats/libbeat/logp"
)

File: x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+// you may not use this file except in compliance with the Elastic License.
+
+package elb
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/provider_test.go]: Nit. This import should be at the end, with the other `libbeat` subpackage.

File: libbeat/docs/shared-autodiscover.asciidoc [hunk 27]
```diff
@@ -257,6 +261,41 @@ the ones used for discovery probes, they have these settings:
 include::../../{beatname_lc}/docs/autodiscover-jolokia-config.asciidoc[]
 endif::autodiscoverJolokia[]
 
+ifdef::autodiscoverAWSELB[]
+[float]
+===== Amazon ELBs
```
Comment (jsoriano) [inline libbeat/docs/shared-autodiscover.asciidoc]: We should probably remark here that this provider is experimental.

File: x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go (hunk 23 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/fetch.go]: Good  call

**Comment by andrewvc** [comment]
@jsoriano I made some material improvements in https://github.com/andrewvc/beats/compare/282cec2...a5340f1ee would you mind giving me one final approval? I found some bugs in manual testing that are now fixed.

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go [hunk 28]
```diff
@@ -66,11 +65,12 @@ func (w *watcher) forever() {
 // once executes the watch loop a single time.
 // This is mostly useful for testing.
 func (w *watcher) once() error {
-	ctx, _ := context.WithTimeout(context.Background(), w.interval)
+	ctx, _ := context.WithTimeout(context.Background(), w.period)
 	fetchedLbls, err := w.fetcher.fetch(ctx)
 	if err != nil {
 		return err
 	}
+	logp.Info("fetched %d load balancer listeners from AWS for autodiscovery", len(fetchedLbls))
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: Maybe logging this on every loop is too much, log this at the debug level?

File: x-pack/libbeat/autodiscover/providers/aws/elb/config.go [hunk 29]
```diff
@@ -30,8 +30,8 @@ type Config struct {
 
 	// AWS Specific autodiscover fields
 
-	Regions []string `config:"regions" validate:"required"`
-	aws.ConfigAWS
+	Regions   []string      `config:"regions" validate:"required"`
+	AWSConfig aws.ConfigAWS `config:",inline"`
```
Comment (jsoriano) [inline x-pack/libbeat/autodiscover/providers/aws/elb/config.go]: Was there any problem with this field embedded?

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go (hunk 28 above)
Comment (jsoriano) [reply x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: Btw, in any case I think this should be `Infof` so it formats the string.
```suggestion
	logp.Infof("fetched %d load balancer listeners from AWS for autodiscovery", len(fetchedLbls))
```

File: x-pack/libbeat/autodiscover/providers/aws/elb/config.go (hunk 29 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/config.go]: Yeah, it wasn't stylistically inline with other uses of the AWS config stuff which keeps it inline.

File: x-pack/libbeat/autodiscover/providers/aws/elb/watch.go (hunk 28 above)
Comment (andrewvc) [reply x-pack/libbeat/autodiscover/providers/aws/elb/watch.go]: Thanks, will convert it to debug thanks!

**Comment by andrewvc** [comment]
Failure is unrelated
